# Retail Sales Data Analysis & Customer Segmentation

## Level 1: Data Understanding

### Objective

The objective of this phase is to gain a comprehensive understanding of the Online Retail dataset before performing any cleaning or analysis.

This includes:

- Understanding the structure and contents of the dataset
- Identifying the purpose of each column
- Assessing data quality through missing value and duplicate analysis
- Understanding customer, product, and geographic coverage
- Examining the time period covered by the data
- Identifying potential data quality issues such as invalid quantities and prices
- Evaluating whether the dataset is suitable for retail sales analysis and customer segmentation

By the end of this phase, a clear understanding of the dataset's strengths, limitations, and preparation requirements will be established for the subsequent data cleaning and analysis stages.

---

### Environment Setup

The following libraries will be used throughout the project:

- Pandas for data handling
- NumPy for numerical operations
- Matplotlib for visualization
- Seaborn for statistical visualization

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

print("Environment Ready")

Environment Ready


All required libraries were imported successfully and the analysis environment is ready.

---

### Dataset Loading

In [3]:
df = pd.read_csv("../data/Online Retail.csv", encoding="latin1")

The dataset was loaded successfully.

Each row represents a retail transaction containing invoice details, product information, customer information, pricing, and country information.

---

### Initial Data Preview

In [4]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12-01-2010 08:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12-01-2010 08:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12-01-2010 08:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12-01-2010 08:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12-01-2010 08:26,3.39,17850.0,United Kingdom


The first five records were displayed successfully.

The dataset contains transactional information including:

- Invoice Number
- Product Code (StockCode)
- Product Description
- Quantity Purchased
- Invoice Date
- Unit Price
- Customer ID
- Customer Country

Each row appears to represent a product-level transaction within a customer invoice.

The data structure is consistent with a retail sales dataset and is suitable for customer, product, and revenue analysis.

---

### Dataset Dimensions Analysis

In [5]:
df.shape

(541909, 8)

The dataset contains 541,909 transaction records and 8 columns.

This indicates a large retail sales dataset suitable for customer and revenue analysis.

---

### Column Names Overview

In [6]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='str')

The dataset contains columns related to invoices, products, quantities, transaction dates, prices, customers, and countries.

These columns provide enough information for sales analysis and customer segmentation.

---

### Business Understanding of Dataset Columns

In [7]:
column_description = {
    "InvoiceNo": "Unique invoice number",
    "StockCode": "Unique product code",
    "Description": "Product description",
    "Quantity": "Number of units purchased",
    "InvoiceDate": "Transaction date and time",
    "UnitPrice": "Price per unit",
    "CustomerID": "Unique customer identifier",
    "Country": "Customer country"
}

column_description

{'InvoiceNo': 'Unique invoice number',
 'StockCode': 'Unique product code',
 'Description': 'Product description',
 'Quantity': 'Number of units purchased',
 'InvoiceDate': 'Transaction date and time',
 'UnitPrice': 'Price per unit',
 'CustomerID': 'Unique customer identifier',
 'Country': 'Customer country'}

Each column represents an important business entity.

InvoiceNo identifies transactions, CustomerID identifies customers, and UnitPrice with Quantity will be used to calculate revenue.

---

### Data Types Analysis

In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  str    
 1   StockCode    541909 non-null  str    
 2   Description  540455 non-null  str    
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  str    
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 33.1 MB


The dataset contains numerical, categorical, and datetime variables.

InvoiceDate has been converted successfully into datetime format for time-series analysis.

---

### Memory Usage Assessment

In [9]:
df.info(memory_usage="deep")

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  str    
 1   StockCode    541909 non-null  str    
 2   Description  540455 non-null  str    
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  str    
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 173.4 MB


The dataset occupies a moderate amount of memory and can be processed efficiently using pandas.

---

### Missing Values Analysis

In [10]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

Missing values exist primarily in CustomerID and Description columns.

These missing values will require treatment during the data cleaning phase.

---

### Duplicate Records Analysis

In [11]:
df.duplicated().sum()

np.int64(5268)

The dataset contains duplicate transaction records.

These duplicates may inflate sales calculations and should be removed during cleaning.

---

### Random Transaction Inspection

In [12]:
df.sample()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
450402,575176,22539,MINI JIGSAW DOLLY GIRL,1,11-08-2011 18:29,0.83,NaN,United Kingdom


The randomly sampled transactions confirm that the dataset contains detailed retail transaction records, including invoice numbers, product codes, product descriptions, quantities, transaction dates, unit prices, customer identifiers, and customer countries.

The sample also reveals that some records contain missing CustomerID values, which supports the earlier missing value analysis.

Overall, the dataset structure appears consistent and suitable for retail sales analysis, customer segmentation, and revenue analysis.

---

### Customer Coverage Analysis

In [13]:
df["CustomerID"].nunique()

4372

The business served 4,372 unique customers during the observed period.

This customer base will be used later for segmentation and CLV analysis.

---

### Product Coverage Analysis

In [14]:
df["StockCode"].nunique()

4070

The dataset contains 4,070 unique products.

This indicates a diverse product catalog.

---

### Geographic Coverage Analysis

In [15]:
df["Country"].nunique()

38

Sales transactions were recorded across 38 countries.

This suggests that the business serves an international customer base.

---

### Date Format Conversion

In [16]:
df["InvoiceDate"] = pd.to_datetime(
    df["InvoiceDate"],
    format="mixed"
)

The InvoiceDate column was successfully converted to datetime format.

This will enable trend analysis, seasonality analysis, and time-based feature engineering later in the project.

---

### Sales Period Analysis

In [17]:
df["InvoiceDate"].min(), df["InvoiceDate"].max()

(Timestamp('2010-12-01 08:26:00'), Timestamp('2011-12-09 12:50:00'))

The dataset covers transactions from 01 December 2010 to 09 December 2011.

This provides approximately one year of retail sales history, which is sufficient for analyzing customer behavior, revenue trends, and seasonal patterns.

---

### Statistical Summary of Numerical Variables

In [18]:
df.describe()

,Quantity,InvoiceDate,UnitPrice,CustomerID
count,541909.000000,541909,541909.000000,406829.000000
mean,9.552250,2011-07-04 13:34:57.156386,4.611114,15287.690570
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000
25%,1.000000,2011-03-28 11:34:00,1.250000,13953.000000
50%,3.000000,2011-07-19 17:17:00,2.080000,15152.000000
75%,10.000000,2011-10-19 11:27:00,4.130000,16791.000000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000
std,218.081158,NaN,96.759853,1713.600303


Summary statistics provide insights into quantity purchased, product pricing, and customer identifiers.

Potential outliers can already be observed and will be investigated further during cleaning.

---

### Invalid Quantity Assessment

In [19]:
(df["Quantity"] <= 0).sum()

np.int64(10624)

There are 10,624 transactions with zero or negative quantities.

These records most likely represent product returns, cancellations, or invalid transactions.

Such records can distort sales and revenue analysis and will be handled during the data cleaning phase.

---

### Invalid Price Assessment

In [20]:
(df["UnitPrice"] <= 0).sum()

np.int64(2517)

There are 2,517 transactions with zero or negative unit prices.

These records may represent free items, promotional products, or data quality issues.

They require further investigation before revenue calculations are performed.

---

### Country-wise Transaction Distribution

In [21]:
df["Country"].value_counts().head(10)

Country
United Kingdom    495478
Germany             9495
France              8557
EIRE                8196
Spain               2533
Netherlands         2371
Belgium             2069
Switzerland         2002
Portugal            1519
Australia           1259
Name: count, dtype: int64

The United Kingdom dominates transaction volume with 495,478 transactions.

Germany, France, EIRE, and Spain are among the next largest markets.

This indicates that the business is heavily dependent on the United Kingdom while maintaining an international customer presence.

---

### Summary

#### Dataset Overview

The Online Retail dataset contains transactional sales records from a UK-based online retailer covering the period from December 2010 to December 2011.

The dataset consists of 541,909 transaction records and 8 business-related attributes, including invoice information, product details, customer information, pricing data, and geographic information.

The dataset provides sufficient information for conducting retail sales analysis, customer behavior analysis, customer lifetime value (CLV) analysis, and customer segmentation.

---

#### Data Quality Assessment

The following data quality issues were identified during the initial assessment:

- Missing values exist in the CustomerID and Description columns.
- 5,268 duplicate records were detected.
- 10,624 transactions contain non-positive quantities.
- 2,517 transactions contain non-positive unit prices.

These issues may affect the accuracy of analytical results and therefore require further cleaning and validation.

---

#### Business Coverage

Key business characteristics identified from the dataset include:

- 4,372 unique customers.
- 4,070 unique products.
- Transactions across 38 countries.
- The United Kingdom represents the largest market by transaction volume.
- More than one year of transactional history is available for analysis.

This level of coverage provides a strong foundation for customer analytics and sales performance evaluation.

---

#### Readiness for Analysis

The dataset contains rich customer, product, pricing, and transaction information. However, several data quality issues must be resolved before conducting exploratory analysis and customer segmentation.

The dataset is suitable for advanced retail analytics once appropriate cleaning and preparation steps have been completed.

---

#### Conclusion

The Data Understanding phase provided a comprehensive overview of the dataset's structure, business context, and overall data quality.

The analysis identified several data quality issues, including missing values, duplicate records, and invalid transaction entries, which require further attention before analytical modeling and business analysis can be performed.

Overall, the dataset contains sufficient customer, product, pricing, and transactional information to support retail sales analysis, customer lifetime value (CLV) analysis, and customer segmentation.

The findings from this phase will serve as the foundation for the next stage of the project: **Data Cleaning & Preparation**.

---

#### Next Step

**Level 2: Data Cleaning & Preparation**

The insights gathered during the Data Understanding phase highlighted several data quality issues that may affect analytical accuracy. The next phase will focus on cleaning and preparing the dataset to ensure a reliable foundation for business analysis and customer segmentation.

